submariner/fine-grained-connectivity.md

```diff
@@ -0,0 +1,86 @@
+# Fine Grained Connectivity
+
+## Summary
+
+Submariner currently provisions full mesh topology by default between the participating clusters.
+For some use-cases, we need fine-grained connectivity options. This proposal covers only the topology
+aspect of connectivity.
+
+This proposal aims to address the below use-case
+
+Hub-Spoke Topology -  Here only the client can talk to the server but not to the other clients to
+which the server is connected to.
+
+To achieve each hub and spoke will be grouped together into an isolated group of clusters called Clusterset.
+There will be a full mesh topology within the Clusterset, though in this use case it will be just two clusters
+in the Clusterset. A cluster can be a part of multiple Clusterset. So the hub can be a part of multiple
+Clustersets, where it connects with a spoke in each Clusterset
+
+## Proposal
+
+### Terminology
+
+#### Clusterset
+
+A Clusterset is a group of clusters that forms a trust boundary. The design of Clusterset can be aligned
+to the MCS KEP. More details are available in the below link.
+
+[Cluster-Id KEP]: http://www.reddit.comhttps://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/2149-clusterid
+
+In this proposal, we intend to use only the ClusterId(which exists already) and ClusterSet as a part of
+Submariner Cluster CRD. The ClusterClaim API shall be realized once the KEP is finalized upstream.
+
+### Design Details
+
+![Clustersets](./images/clustersets.png)
+
+The Submariner should be enhanced to support grouping of clusters into Clustersets. The Cluster CRD can
+be enhanced to hold the Clustersets name in which the cluster is a part of. The one or more Clustersets
+the cluster wants to be a part of shall be specified when the cluster joins the broker. If no
+Clustersets are specified, the default will be used. The cluster-id should be unique across Clustersets.
+
+```Go
+
+type Cluster struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Spec              ClusterSpec `json:"spec"`
+}
+
+type ClusterSpec struct {
+        ClusterID   string   `json:"cluster_id"` // perhaps this could just be a hash of the name...?
+        ClusterSets []string `json:"cluster_sets"`
+        ServiceCIDR []string `json:"service_cidr"`
+        ClusterCIDR []string `json:"cluster_cidr"`
+        GlobalCIDR  []string `json:"global_cidr"`
+}
+```
+
+#### Submariner Broker
+
+There will be a single broker running in one of the participating clusters. The clusters will specify
+the Clusterset it wants to be a part of when it joins the broker. A cluster can be a part of multiple Clustersets.
+The broker ensures that resources are only synced within a Clusterset.
```

Comment:
A broker is basically a namespace hosted on an API server. I think it makes sense for each cluster set to have its own broker namespace so they don't conflict so, in terms of the broker, a cluster set equates to a namespace. So logically there would be multiple brokers, one per cluster set. Otherwise, with a single broker namespace, we'd need some other way to route resources to the correct clusters (probably annotate the resources with the cluster set ID). The brokers could be hosted on one API server or multiple, that would be up to the admin.

Reply:
I agree that Broker is just a namespace today. Inorder to support multiple Brokers for each Clusterset, we will have to modify ClusterCRD to include Broker URL/Credentials for each ClusterSet. This way, we will have Single Submariner Components (which will avoid race conditions while applying rules on the nodes) and will be able to support multiple Brokers. But this will be a bit intrusive change and also LH might need awareness of Clusetset (and the associated Broker in the ClusterSet).

Reply:
yes each component that syncs to/from the broker would need to be aware of each broker URL, broker namespace and local namespace. Of course today we pass that via env vars but could be `ConfigMap`(s). It doesn't seem like submariner would need to be aware of the concept of a Clusterset - that seems like a higher-level logical organization.

Reply:
@tpantelis  As discussed we can have singe broker and single broker n/s and use the clusterset/colorcode in the clusterCRD to identify how the cluster needs to be grouped together. Lighthouse will have to add the clusterset/colorcode as a label in ServiceImport. So the high-level design will remain more or less the same as described in the proposal. I will add more details about the design, once we decide to go ahead with this proposal.